### PR TITLE
add std.os.shutdown function for sockets

### PR DESCRIPTION
**Pull request by marler8997**

### CODE REVIEW

**Comment by LemonBoy** [comment]
You may be interested in #7194 and the sibling #7124

File: lib/std/os/test.zig [hunk 1]
```diff
@@ -627,3 +627,19 @@ test "getrlimit and setrlimit" {
         try os.setrlimit(resource, limit);
     }
 }
+
+test "shutdown socket" {
+    if (builtin.os.tag != .linux and builtin.os.tag != .windows)
```
Comment (LemonBoy) [inline lib/std/os/test.zig:632]: NO.

Please fill the entries for [openbsd](https://github.com/openbsd/src/blob/master/sys/sys/socket.h), [netbsd](http://fxr.watson.org/fxr/source/sys/socket.h?v=NETBSD) and [freebsd](http://fxr.watson.org/fxr/source/sys/socket.h).

File: lib/std/os/linux.zig [hunk 2]
```diff
@@ -1005,7 +1005,7 @@ pub fn recvfrom(fd: i32, noalias buf: [*]u8, len: usize, flags: u32, noalias add
     return syscall6(.recvfrom, @bitCast(usize, @as(isize, fd)), @ptrToInt(buf), len, flags, @ptrToInt(addr), @ptrToInt(alen));
 }
 
-pub fn shutdown(fd: i32, how: i32) usize {
+pub fn shutdown(fd: i32, how: c_int) usize {
```
Comment (LemonBoy) [inline lib/std/os/linux.zig:1008]: Native zig types are preferred, why this change?

File: lib/std/os.zig [hunk 3]
```diff
@@ -2706,6 +2706,58 @@ pub fn socket(domain: u32, socket_type: u32, protocol: u32) SocketError!socket_t
...
+    /// The network subsystem has failed.
+    NetworkSubsystemFailed,
+
+    /// The socket is not connected (connection-oriented sockets only).
+    SocketNotConnected,
+
+    SystemResources
+} || UnexpectedError;
+
+pub const ShutdownHow = enum {
```
Comment (LemonBoy) [inline lib/std/os.zig:2726]: I'd make this a simple Zig enum `enum { read_end, write_end, both_ends }` and convert this to the proper flag value in `fn shutdown`.

File: lib/std/net.zig [hunk 4]
```diff
@@ -912,6 +912,7 @@ fn linuxLookupName(
             os.connect(fd, da, dalen) catch break :syscalls;
             key |= DAS_USABLE;
             os.getsockname(fd, sa, &salen) catch break :syscalls;
+            os.shutdown(fd, .both) catch {};
```
Comment (LemonBoy) [inline lib/std/net.zig:915]: The `closeSocket` call is enough here, this is an UDP socket.

File: lib/std/os.zig (hunk 3 above)
Comment (marler8997) [reply lib/std/os.zig:2726]: I dunno, this seems simpler than adding conversion logic.

File: lib/std/os/test.zig (hunk 1 above)
Comment (marler8997) [reply lib/std/os/test.zig]: ok I've filled out the entries for those platforms as well

File: lib/std/os/linux.zig (hunk 2 above)
Comment (marler8997) [reply lib/std/os/linux.zig]: ok I've removed this change and modified the new declaration in `c.zig` to use `i32`.

File: lib/std/c.zig [hunk 5]
```diff
@@ -145,6 +145,7 @@ pub extern "c" fn ioctl(fd: fd_t, request: c_int, ...) c_int;
 pub extern "c" fn uname(buf: *utsname) c_int;
 
 pub extern "c" fn gethostname(name: [*]u8, len: usize) c_int;
+pub extern "c" fn shutdown(socket: fd_t, how: i32) usize;
```
Comment (semarie) [inline lib/std/c.zig:148]: for `how`, I would use a `c_int` instead of `i32`. `c_int` could have different size depending the platform.

Comment (semarie) [reply lib/std/c.zig:148]: oh, I see @LemonBoy said the opposite before for `lib/std/os/linux.zig` function.

@LemonBoy, using `c_int` ensure portability: C standard (on which `shutdown()` POSIX is based) defines no size of `int` (only a minimum range, and relative size compared to `short` and `long`)

Comment (semarie) [reply lib/std/c.zig:148]: and I think also the return value is wrong: `usize` is unsigned, whereas POSIX `shutdown()` could return `-1` on error. I would use `c_int` here too :smile:

File: lib/std/os.zig (hunk 3 above)
Comment (LemonBoy) [reply lib/std/os.zig:2726]: It's a leaky abstraction. The "conversion" code is a simple switch, the same you're doing here with a lot of `if`s.

File: lib/std/c.zig (hunk 5 above)
Comment (LemonBoy) [reply lib/std/c.zig:148]: > oh, I see @LemonBoy said the opposite before for lib/std/os/linux.zig function.

That's the Linux syscall wrapper, we're directly talking with the kernel there and don't give a damn about C/libc, the types there should be Zig's native ones.

The `c.zig` definitions, on the other hand, should use the C types instead to guarantee the ABI compatibility.

> and I think also the return value is wrong: usize is unsigned, whereas POSIX shutdown() could return -1 on error. I would use c_int here too

Good catch, `c_int` is the correct return type.

File: lib/std/os.zig (hunk 3 above)
Comment (marler8997) [reply lib/std/os.zig]: ok I've changed it to switches internal to the function.  This seems like it could be adding unnecessary overhead to me, but not a big deal.

File: lib/std/c.zig (hunk 5 above)
Comment (marler8997) [reply lib/std/c.zig]: Ok I've modified the one in `c.zig` to:
```c
pub extern "c" fn shutdown(socket: fd_t, how: c_int) c_int;
```

File: lib/std/os/test.zig [hunk 6]
```diff
@@ -627,3 +627,19 @@ test "getrlimit and setrlimit" {
...
+    if (builtin.os.tag == .windows) {
+        _ = try std.os.windows.WSAStartup(2, 2);
+    }
+    defer {
+        if (builtin.os.tag == .windows) {
+            std.os.windows.WSACleanup() catch unreachable;
+        }
+    }
+    const sock = try os.socket(os.AF_INET, os.SOCK_STREAM, 0);
+    os.shutdown(sock, .both) catch {}; // should fail
```
Comment (LemonBoy) [inline lib/std/os/test.zig:643]:
```suggestion
    os.shutdown(sock, .both) catch |err| switch (err) {
        error.SocketNotConnected => {},
        else => |e| return e,
    }
```
